### DISCUSSION TITLE
Differences in Pyswmm and SWMM time series results

### DISCUSSION THREAD

**Discussion by zsbeheshti** [pyswmm/pyswmm#391] [Q&A] [answered]
Dear all,

With the help of Pyswmm, I change some objects’ features (e.g. pipe conduits diameters) and then run the Pyswmm. After the run, I extract the results of Heads and Flows using Pyswmm’s output functions. However, the results of the time series aren’t the same as the corresponding results with the SWMM run from outside Pyswmm, and the differences are significant! Does anyone have any related experience to share with me?

 

Regards

**Comment by bemcdonnell** [comment]
@zsbeheshti it is very difficult to help you if you do not provide example code and models for us to try to run.  Can you please share?  Moreover, please use the "discussion" section rather than "issue."  This is not a bug until we confirm the process. Does that makes sense?

**Reply by zsbeheshti** [reply]
Yes, that's right. Thanks.

With the help of Pyswmm, I change some objects' features, and I extract the results of Flows using Pyswmm's output functions:

from pyswmm import Simulation, Links
with Simulation('3tanks.inp') as sim:
    C1 = Links(sim)["C-1"]
    C1.inlet_offset=3
    sim.step_advance(890)
    for step in sim:
          print(sim.current_time)
          sim.report()
          print(C1.flow)

SWMM:

Table - Link C-1 |  Flow

Days | Hours | (CMS)
0 | 0:15:00 | 1.9
0 | 0:30:00 | 2.01
0 | 0:45:00 | 2
0 | 1:00:00 | 2
0 | 1:15:00 | 2
0 | 1:30:00 | 2
0 | 1:45:00 | 2
0 | 2:00:00 | 2
0 | 2:15:00 | 2
0 | 2:30:00 | 2
0 | 2:45:00 | 2
0 | 3:00:00 | 2
0 | 3:15:00 | 2
0 | 3:30:00 | 2
0 | 3:45:00 | 2.04
0 | 4:00:00 | 2.51
0 | 4:15:00 | 4.63
0 | 4:30:00 | 7.81
0 | 4:45:00 | 9.66
0 | 5:00:00 | 11.1
0 | 5:15:00 | 10.21
0 | 5:30:00 | 8.27
0 | 5:45:00 | 6.63
0 | 6:00:00 | 5.51
0 | 6:15:00 | 4.74
0 | 6:30:00 | 4.18
0 | 6:45:00 | 3.77
0 | 7:00:00 | 3.46
0 | 7:15:00 | 3.22
0 | 7:30:00 | 3.03
0 | 7:45:00 | 2.88
0 | 8:00:00 | 2.75
0 | 8:15:00 | 2.65
0 | 8:30:00 | 2.57
0 | 8:45:00 | 2.5
0 | 9:00:00 | 2.44
0 | 9:15:00 | 2.39
0 | 9:30:00 | 2.34
0 | 9:45:00 | 2.3
0 | 10:00:00 | 2.27
0 | 10:15:00 | 2.24
0 | 10:30:00 | 2.22
0 | 10:45:00 | 2.2
0 | 11:00:00 | 2.18
0 | 11:15:00 | 2.17
0 | 11:30:00 | 2.15
0 | 11:45:00 | 2.14
0 | 12:00:00 | 2.13
0 | 12:15:00 | 2.12
0 | 12:30:00 | 2.11
0 | 12:45:00 | 2.1
0 | 13:00:00 | 2.09
0 | 13:15:00 | 2.09
0 | 13:30:00 | 2.08
0 | 13:45:00 | 2.07
0 | 14:00:00 | 2.07
0 | 14:15:00 | 2.07
0 | 14:30:00 | 2.06
0 | 14:45:00 | 2.06
0 | 15:00:00 | 2.06
0 | 15:15:00 | 2.05
0 | 15:30:00 | 2.05
0 | 15:45:00 | 2.05
0 | 16:00:00 | 2.05
0 | 16:15:00 | 2.04
0 | 16:30:00 | 2.04
0 | 16:45:00 | 2.04
0 | 17:00:00 | 2.04
0 | 17:15:00 | 2.04
0 | 17:30:00 | 2.03
0 | 17:45:00 | 2.03



Pyswmm:


2014-05-03 00:15:04
1.9008743142287912
2014-05-03 00:30:04
2.081977863138803
2014-05-03 00:45:04
2.397771086874463
2014-05-03 01:00:04
1.8015071914637235
2014-05-03 01:15:04
1.846404056647297
2014-05-03 01:30:04
1.8711989688684878
2014-05-03 01:45:04
1.873310484802351
2014-05-03 02:00:04
1.875157751982357
2014-05-03 02:15:04
2.6194087119481386
2014-05-03 02:30:04
1.87619360423507
2014-05-03 02:45:04
1.8754175832916522
2014-05-03 03:00:04
2.6195246298002166
2014-05-03 03:15:04
1.8761875691642462
2014-05-03 03:30:04
1.875415594850679
2014-05-03 03:45:04
1.8915501135691095
2014-05-03 04:00:04
2.9165153563261548
2014-05-03 04:15:04
5.29284895779272
2014-05-03 04:30:04
7.102950759753539
2014-05-03 04:45:04
9.506554961540031
2014-05-03 05:00:04
10.99412114855903
2014-05-03 05:15:04
10.338832637773963
2014-05-03 05:30:04
8.464891111857629
2014-05-03 05:45:04
6.7933051830467095
2014-05-03 06:00:04
5.631129844435614
2014-05-03 06:15:04
4.8263100927406875
2014-05-03 06:30:04
4.249512502116297
2014-05-03 06:45:04
3.8232699923072206
2014-05-03 07:00:04
3.5005673206944676
2014-05-03 07:15:04
3.251023510175582
2014-05-03 07:30:04
3.054600835002495
2014-05-03 07:45:04
2.8975696141814793
2014-05-03 08:00:04
2.7700828000040474
2014-05-03 08:15:04
2.6654584866835704
2014-05-03 08:30:04
2.5787114607328756
2014-05-03 08:45:04
2.5061363057678885
2014-05-03 09:00:04
2.4449498070773346
2014-05-03 09:15:04
2.3930460558997924
2014-05-03 09:30:04
2.3487048103999606
2014-05-03 09:45:04
2.310597430575575
2014-05-03 10:00:04
2.2777240391803804
2014-05-03 10:15:04
2.2493254916270313
2014-05-03 10:30:04
2.2248560959894133
2014-05-03 10:45:04
2.203511220534206
2014-05-03 11:00:04
2.149678674541877
2014-05-03 11:15:04
2.92967970118635
2014-05-03 11:30:04
2.084298429528438
2014-05-03 11:45:04
2.0638190000274395
2014-05-03 12:00:04
2.0341221227860617
2014-05-03 12:15:04
2.8109033357044124
2014-05-03 12:30:04
2.7955418182072784
2014-05-03 12:45:04
1.9961195432794359
2014-05-03 13:00:04
1.991367550317317
2014-05-03 13:15:04
1.972470223995983
2014-05-03 13:30:04
1.9653765285927203
2014-05-03 13:45:04
1.9605514855457846
2014-05-03 14:00:04
1.9562982674375229
2014-05-03 14:15:04
2.710023358247108
2014-05-03 14:30:04
1.9411347394515164
2014-05-03 14:45:04
1.9409095359850472
2014-05-03 15:00:04
1.93121261184695
2014-05-03 15:15:04
1.937556711620594
2014-05-03 15:30:04
1.9197759175251738
2014-05-03 15:45:04
1.916985019743961
2014-05-03 16:00:04
1.9280005974265226
2014-05-03 16:15:04
1.9126504804680198
2014-05-03 16:30:04
1.9105497338801343
2014-05-03 16:45:04
1.9219702172865754
2014-05-03 17:00:04
1.9070651436778592
2014-05-03 17:15:04
1.905328563178064
2014-05-03 17:30:04
1.9170444035592953
2014-05-03 17:45:04
1.9024851188182574

**Comment by dickinsonre** [comment]
I would add a few questions:
1. What version of SWMM are you using?  Is it newer than the engine in PySWMM?
2. The first thought in my head when I hear users say this is that some of the simulation parameters are different when you moved the inp file to SWMM5.x Time Step etc

**Reply by zsbeheshti** [reply]
swmm version is 5.1

**Comment by dickinsonre** [comment]
It is a bit hard to see the PySWMM numbers but they seem to be off by 5 percent  Can you attach your SWMM5 file and there are about 15 versions of SWMM 5.1  so saying 5.1 is not precise to say the least.

**Reply by bemcdonnell** [reply]
@dickinsonre I think OWA SWMM is on SWMM-5.1.14

**Comment by bemcdonnell** [comment, marked as answer]
@zsbeheshti, I have an explanation for this.  As a SWMM simulation steps forward the "reporting steps" do not perfectly land on the "routing steps."  SWMM interpolates the output if a report step is between two routing steps.  See the example for "nodes"

https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/7e7e6da799e4c8f1fd51499cebe47f82f80c67f6/src/solver/output.c#L630-L656

As pyswmm runs and you scrape results out, you are actually pulling values directly from the routing steps. This is why "step advance" is useful.   If you want to test this, turn off step advance, get ALL the routing step values for the flow rate and then interpolate the values yourself.  I bet the interpolated numbers will be exactly the same between PySWMM and the OUT file.

**Reply by bemcdonnell** [reply]
@zsbeheshti, another note about your code `sim.report()` should only be called After your simulation, not during
```python
from pyswmm import Simulation, Links
with Simulation('3tanks.inp') as sim:
    C1 = Links(sim)["C-1"]
    C1.inlet_offset=3
    sim.step_advance(890)
    for step in sim:
        print(sim.current_time)
        print(C1.flow)
    sim.report()
```

**Reply by bemcdonnell** [reply]
@zsbeheshti did I answer your question?

**Reply by zsbeheshti** [reply]
Yes, Thanks.

**Comment by zsbeheshti** [comment]
@bemcdonnell I checked it. Thank you for your guidance.
